02-react-week-1/spreading-and-destructuring.md

```diff
@@ -0,0 +1,379 @@
+
+```javascript
+const letters = [...'abc']; // ['a', 'b', 'c']
+```
+
+### Combining the spread operator with destructuring
+
+You can also use destructuring along with the spread operator to remove individual elements from an array.
+
+```javascript
+const list = ['peanut butter', 'chocolate', 'vanilla', 'cookies and cream'];
+
+const [first, second, ...rest] = list;
+
+console.log(first); // peanut butter
+console.log(second); // chocolate
+console.log(rest); // vanilla,cookies and cream 
+```
+
+### Spread operator with object literals
+
+The spread operator can also be used to combine or merge two object literals.  The code below merges two object literals.
+
+```javascript
+const ada = {
+  name: 'Ada Lovelace',
+  age: 103,
+};
+
+const language = {
+  name: 'C#',
+  about: 'A strongly typed language by Microsoft',
+};
+
+const merged = { ...ada, ...language };
+
+console.log(merged); // { name: 'C#', age: 103, about: 'A strongly typed language by Microsoft' }
+```
+
+Notice that it merges the objects in order, so that `ada`'s name is overwritten by that of `language`.  It also creates an entirely **new** array which shallowly copies elements of the original object.  This means that sub-objects or arrays can be references to the original.
+
+```javascript
+const student = {
+  name: 'Bob',
+  grades: [3, 4, 5],
+  address: {
+    street: '123 Elm St',
+    city: 'Trenton',
+    state: 'NJ',
+    zip: 11111,
+   }
+};
+
+const newStudent = { ...student };
+
+newStudent.name = 'Alice';
+newStudent.address.street = '4th avenue';
+
+console.log(newStudent);
+console.log('---');
+console.log(student);
+```
+
+The above code outputs: 
+
+```javascript
+{
+  name: 'Alice',
+  grades: [ 12, 4, 5 ],
+  address: {
+    street: '4th avenue',
+    city: 'Trenton',
+    state: 'NJ',
+    zip: 11111
+  }
+}
+---
+{
+  name: 'Bob',
+  grades: [ 12, 4, 5 ],
+  address:
+   { 
+     street: '4th avenue',
+     city: 'Trenton',
+     state: 'NJ',
+     zip: 11111
+  } 
+}
+```
+<br/><br/>
+
+**Question** Look at the code above.  Why do both Alice & Bob have a street address of 4th avenue?
+
+<details>
+  <summary>
+    Click here to see the answer.
+  </summary>
+
+  The `...` operator is making a _shallow_ copy of the original object.  It does not copy all the values of `student`'s properties.  It instead copies the _reference_ to the `address` object and the `grades` array.  To make a _deep copy_ of the original object you would need to do things [differently.](https://medium.com/@tkssharma/objects-in-javascript-object-assign-deep-copy-64106c9aefab)
+</details>
+
+
+## Examining Code
+
+The following code simulates merging new permissions into a user account.  With a partner examine the code and answer the questions:
+
+```javascript
+const combinePreferences = function combinePreferences(user, ...preferences) {
+  preferences.forEach((preference) => {
+    user.preferences = { ...user.preferences, ...preference}
+  });
+  return user;
+}
```

Comment:
the rest param variable name seems confusing in this context; however, if this was an intentional variable name then ignore this
```suggestion
const combinePreferences = function combinePreferences(user, ...otherPrefs) {
  otherPrefs.forEach((preference) => {
    user.preferences = { ...user.preferences, ...preference}
  });
  return user;
}
```

Reply:
you might need to expand the code to see the entire context